File: src/harness/adapters.ts
/**
 * Harness adapters — one per supported coding harness (spec: Claude Code,
 * Codex, Antigravity, Cursor, Hermes, Pi, OpenCode).
 *
 * Adapters are declarative: instruction file, detection signals, capability
 * flags, and an optional per-harness rendering of the canonical instructions
 * (Cursor needs `.cursor/rules/*.mdc` frontmatter; plain-markdown harnesses
 * share the managed-block rendering).
 *
 * Several harnesses read the open-standard AGENTS.md; translation dedupes by
 * instruction file path so shared files are written once.
 */
import type { HarnessAdapter } from "../types.ts";

export interface RenderedInstructionFile {
  /** Repo-relative path to write. */
  path: string;
  /** Prefix written before the managed block (e.g. .mdc frontmatter) when the file is created fresh. */
  freshFilePrefix: string;
}

export const HARNESS_ADAPTERS: HarnessAdapter[] = [
  {
    id: "claude-code",
    title: "Claude Code",
    instructionFile: "CLAUDE.md",
    cliNames: ["claude"],
    configSignals: ["CLAUDE.md", ".claude/settings.json", ".claude"],
    capabilities: { hooks: true, mcp: true, permissions: true },
  },
  {
    id: "codex",
    title: "Codex",
    instructionFile: "AGENTS.md",
    cliNames: ["codex"],
    configSignals: ["AGENTS.md", ".codex"],
    capabilities: { hooks: false, mcp: true, permissions: false },
  },
  {
    id: "cursor",
    title: "Cursor",
    instructionFile: ".cursor/rules/ade.mdc",
    cliNames: ["cursor"],
    // .cursorrules is the legacy single-file surface — a detection signal, not a write target.
    configSignals: [".cursor/rules", ".cursorrules", ".cursor"],
    capabilities: { hooks: false, mcp: true, permissions: false },
  },
  {
    id: "opencode",
    title: "OpenCode",
    instructionFile: "AGENTS.md",
    cliNames: ["opencode"],
    configSignals: ["AGENTS.md", "opencode.json"],
    capabilities: { hooks: false, mcp: true, permissions: true },
  },
  {
    id: "antigravity",
    title: "Antigravity",
    instructionFile: "AGENTS.md",
    cliNames: ["antigravity"],
    configSignals: ["AGENTS.md", ".antigravity"],
    capabilities: { hooks: false, mcp: false, permissions: false },
  },
  {
    id: "hermes",
    title: "Hermes",
    instructionFile: "AGENTS.md",
    cliNames: ["hermes"],
    configSignals: ["AGENTS.md", ".hermes"],
    capabilities: { hooks: false, mcp: false, permissions: false },
  },
  {
    id: "pi",
    title: "Pi",
    instructionFile: "AGENTS.md",
    cliNames: ["pi"],
    configSignals: ["AGENTS.md", ".pi"],
    capabilities: { hooks: false, mcp: false, permissions: false },
  },
];

export function getAdapter(id: string): HarnessAdapter | undefined {
  return HARNESS_ADAPTERS.find((adapter) => adapter.id === id);
}

/**
 * Per-harness rendering details for the canonical instructions.
 * Cursor project rules are `.mdc` with frontmatter; everything else is a
 * plain markdown managed block.
 */
export function renderTarget(adapter: HarnessAdapter): RenderedInstructionFile {
  if (adapter.id === "cursor") {
    return {
      path: adapter.instructionFile,
      freshFilePrefix: "---\ndescription: ADE Bootstrapper baseline (managed)\nalwaysApply: true\n---\n\n",
    };
  }
  return { path: adapter.instructionFile, freshFilePrefix: "" };
}

File: crates/ade-core/src/harness/adapters.rs
//! Harness adapter registry — port of `src/harness/adapters.ts`.
//! Seven harnesses, capability flags, instruction-file targets.

use crate::types::{HarnessAdapter, HarnessCapabilities};

pub const HARNESS_ADAPTERS: [HarnessAdapter; 7] = [
    HarnessAdapter {
        id: "claude-code",
        title: "Claude Code",
        instruction_file: "CLAUDE.md",
        cli_names: &["claude"],
        config_signals: &["CLAUDE.md", ".claude/settings.json", ".claude"],
        capabilities: HarnessCapabilities {
            hooks: true,
            mcp: true,
            permissions: true,
        },
    },
    HarnessAdapter {
        id: "codex",
        title: "Codex",
        instruction_file: "AGENTS.md",
        cli_names: &["codex"],
        config_signals: &["AGENTS.md", ".codex"],
        capabilities: HarnessCapabilities {
            hooks: false,
            mcp: true,
            permissions: false,
        },
    },
    HarnessAdapter {
        id: "cursor",
        title: "Cursor",
        instruction_file: ".cursor/rules/ade.mdc",
        cli_names: &["cursor"],
        config_signals: &[".cursor/rules", ".cursorrules", ".cursor"],
        capabilities: HarnessCapabilities {
            hooks: false,
            mcp: true,
            permissions: false,
        },
    },
    HarnessAdapter {
        id: "opencode",
        title: "OpenCode",
        instruction_file: "AGENTS.md",
        cli_names: &["opencode"],
        config_signals: &["AGENTS.md", "opencode.json"],
        capabilities: HarnessCapabilities {
            hooks: false,
            mcp: true,
            permissions: true,
        },
    },
    HarnessAdapter {
        id: "antigravity",
        title: "Antigravity",
        instruction_file: "AGENTS.md",
        cli_names: &["antigravity"],
        config_signals: &["AGENTS.md", ".antigravity"],
        capabilities: HarnessCapabilities {
            hooks: false,
            mcp: false,
            permissions: false,
        },
    },
    HarnessAdapter {
        id: "hermes",
        title: "Hermes",
        instruction_file: "AGENTS.md",
        cli_names: &["hermes"],
        config_signals: &["AGENTS.md", ".hermes"],
        capabilities: HarnessCapabilities {
            hooks: false,
            mcp: false,
            permissions: false,
        },
    },
    HarnessAdapter {
        id: "pi",
        title: "Pi",
        instruction_file: "AGENTS.md",
        cli_names: &["pi"],
        config_signals: &["AGENTS.md", ".pi"],
        capabilities: HarnessCapabilities {
            hooks: false,
            mcp: false,
            permissions: false,
        },
    },
];

pub fn get_adapter(id: &str) -> Option<&'static HarnessAdapter> {
    HARNESS_ADAPTERS.iter().find(|adapter| adapter.id == id)
}

pub struct RenderedInstructionFile {
    pub path: &'static str,
    pub fresh_file_prefix: &'static str,
}

/// Per-harness render target: cursor gets `.mdc` frontmatter on fresh files.
pub fn render_target(adapter: &HarnessAdapter) -> RenderedInstructionFile {
    if adapter.id == "cursor" {
        RenderedInstructionFile {
            path: adapter.instruction_file,
            fresh_file_prefix:
                "---\ndescription: ADE Bootstrapper baseline (managed)\nalwaysApply: true\n---\n\n",
        }
    } else {
        RenderedInstructionFile {
            path: adapter.instruction_file,
            fresh_file_prefix: "",
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn seven_adapters_with_unique_ids() {
        assert_eq!(HARNESS_ADAPTERS.len(), 7);
        let mut ids: Vec<&str> = HARNESS_ADAPTERS.iter().map(|adapter| adapter.id).collect();
        ids.sort_unstable();
        ids.dedup();
        assert_eq!(ids.len(), 7);
        assert!(get_adapter("claude-code").is_some());
        assert!(get_adapter("emacs").is_none());
    }

    #[test]
    fn cursor_gets_mdc_frontmatter_others_do_not() {
        let cursor = render_target(get_adapter("cursor").unwrap());
        assert!(cursor.fresh_file_prefix.starts_with("---\n"));
        assert_eq!(cursor.path, ".cursor/rules/ade.mdc");
        let codex = render_target(get_adapter("codex").unwrap());
        assert_eq!(codex.fresh_file_prefix, "");
        assert_eq!(codex.path, "AGENTS.md");
    }

    #[test]
    fn only_claude_code_has_hooks_and_permissions() {
        let hooks: Vec<&str> = HARNESS_ADAPTERS
            .iter()
            .filter(|adapter| adapter.capabilities.hooks)
            .map(|adapter| adapter.id)
            .collect();
        assert_eq!(hooks, vec!["claude-code"]);
    }
}
